gateway/gateway.go

```diff
@@ -1479,6 +1471,39 @@ func (gateway *HandleT) StartAdminHandler(ctx context.Context) error {
 	return g.Wait()
 }
 
+func (gateway *HandleT) maxActiveClientMiddleware(next http.Handler) http.Handler {
+
+	//this is to make sure that we don't have more than `maxClient` in-memory at any point of time. As, having more http client than `maxClient`
+	// may lead to gateway OOM kill.
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		atomic.AddInt32(&gateway.activeClientCount, 1)
+		defer func() {
+			atomic.AddInt32(&gateway.activeClientCount, -1)
+		}()
+		if maxActiveClients != 0 {
+			select {
+			case gateway.activeClient <- struct{}{}:
+				defer func() {
+					<-gateway.activeClient
+				}()
+			default:
+				http.Error(w, response.GetStatus(response.TooManyRequestsInMemory), response.GetStatusCode(response.TooManyRequestsInMemory))
+				return
+			}
+		}
+		next.ServeHTTP(w, r)
+	})
+}
+
+func (gateway *HandleT) statMiddleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```

Comment:
Similar to the above we can avoid coupling with gateway, the only dependency is the stats package. Number of active requests can be counted here.